lib/lattice_observer/observed/event_processor.ex

```diff
@@ -193,33 +195,63 @@ defmodule LatticeObserver.Observed.EventProcessor do
   def record_heartbeat(l = %Lattice{}, source_host, stamp, data) do
     labels = Map.get(data, "labels", %{})
     friendly_name = Map.get(data, "friendly_name", "")
+    annotations = Map.get(data, "annotations", %{})
+    spec = Map.get(annotations, @annotation_app_spec, "")
+
+    # Heartbeats are now considered authoritative, so the previously stored
+    # actor and provider list are wiped prior to recording the heartbeat, but the
+    # host (and its "first seen" time) are not.
+    l =
+      %Lattice{
+        l
+        | actors:
+            l.actors
+            |> Enum.map(fn {k, v} ->
+              {k,
+               %Actor{
+                 v
+                 | instances: v.instances |> Enum.reject(fn i -> i.host_id == source_host end)
+               }}
+            end)
+            |> Enum.into(%{}),
```

Comment:
Let's say there's only one host in the lattice, and the most recent heartbeat has 0 instances of actor X. This will filter out all the instances for the actor. Should the actor entry also be filtered out of the lattice state, or is it okay to leave it with zero instances?

Reply:
In theory an actor with 0 instances shouldn't show up in the heartbeat, but I don't know of the edge case that would make that happen

Reply:
I think I worded the above poorly. I’m describing a case where the lattice monitor’s state contains an actor `A` with 1 instance on host `H`, and the incoming heartbeat data for host `H` contains 0 instances of actor `A`

The `Enum.reject` will filter out the single instance associated with `H`, leaving an empty list. Nothing will be added in the `List.foldl` below, since the host isn’t running an instance of actor `A` anymore

I’m wondering if it’s okay that the new lattice state still contains a reference to actor `A` even though there are 0 instances of it running

Reply:
If there are no instances the actor gets removed as part of the strip instanceless entities call. Unless I'm misunderstanding the question

Reply:
Hosts also don't publish actors with zero instances in a heartbeat

Reply:
Ah, I missed the strip instanceless function 👍